### ISSUE TITLE
Make ICollection members non-explicit on CmsRecipientCollection

### ISSUE THREAD

**Issue by MaximLipnin** [dotnet/corefx#29708]
Hi corefx team,

`CmsRecipientCollection` type has [explicit implementation](https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsRecipientCollection.cs#L115-L129) for ICollection members. 
Shouldn't they be made non-explicit?

Thanks!

**Comment by karelz** [comment]
What is the benefit of them to be non-explicit? The type is sealed ...

**Comment by bartonjs** [comment]
@MaximLipnin Have you finished trying to pull the code over to Mono?  It would be nice to not have a bug a week trickle.

**Comment by MaximLipnin** [comment]
@karelz Just elimination of yet another potential [API break](https://jenkins.mono-project.com/job/test-mono-pull-request-amd64/14325//Public_API_Diff/) in Mono.

@bartonjs No, I haven't. CoreFX import on System.Security doesn't look like a quick jump from one sources to another ones. So I prefer to ask about all non-obvious changes in API (and possible other things). And thanks for your attention and quick response.

**Comment by marek-safar** [comment]
This is a breaking change when coming from .NET framework as you need to cast to the interface to call the members.

@karelz @bartonjs Would you guys really prefer one jumbo issue with all sort of issues instead of separate issue items?

**Comment by marek-safar** [comment]
Well, we'll try to do the simple ones in one go next time. Even if for libraries like mscorlib.dll or System.dll this is not realistic. However, any harder issues will come up later as I don't want to be blocked to collect all problems for weeks (or even months for large assemblies) and then be blocked by you figuring out all the hard one by one.

**Comment by stephentoub** [comment]
> What is the benefit of them to be non-explicit? The type is sealed

@karelz, implicit vs explicit interface implementation impacts whether you can call the method without first casting to the interface.  If I have:
```C#
interface I
{
    void M();
}

class A : I
{
    public void M() { }
}

class B : I
{
    void I.M() { }
}
```
both `A` and `B` implement `I`, and I can do `A a = new A(); a.M();`, but `B b = new B(); b.M();` will not compile as `M` isn't a publicly exposed method of B; I instead have to do `B b = new B(); ((I)b).M();`.  Thus, it's a breaking change to make an interface implementation explicit when it used to be implicit.

**Comment by karelz** [comment]
I assumed we decided to not take the public methods to .NET Core 1.0 and provided just the interface implementation instead.
It looks like we missed the difference during .NET Core 2.0 push towards compat with Framework/Mono. Is that the key problem here?

**Comment by stephentoub** [comment]
Yes

**Comment by danmosemsft** [comment]
I would expect a break of this kind to show up in https://github.com/dotnet/corefx/blob/master/src/shims/ApiCompatBaseline.netcoreapp.netfx461.txt but it does not. @weshaggard seems we have a flaw in apicompat do you agree?

The difference shows up clearly in apisof.net:
https://apisof.net/catalog/System.Security.Cryptography.Pkcs.CmsRecipientCollection.System.Collections.ICollection.SyncRoot
https://apisof.net/catalog/System.Security.Cryptography.Pkcs.CmsRecipientCollection.SyncRoot

**Comment by danmosemsft** [comment]
@bartonjs oh of course -- I guess I was looking for "Interface was implicit but is now only partially implicit" but of course that's a bit unreasonable to expect 😛 

I have gone through this file several times in the past, and should have followed up on these.

Do you plan to fix all these?

**Comment by bartonjs** [comment]
I plan that all of these should be fixed, yes.

I think it's just RecipientInfoCollection and CmsRecipientCollection needing to un-explicit their IsSynchronized and SyncRoot, and the already existing issue for AlgorithmIdentifier.Parameters (#29706).

Everything else in 80-94 (and 74-75) should already be done, and should be removed from the baselines file.

**Comment by danmosemsft** [comment]
Perhaps you could update the baseline with your PR? I'm gratified to see there are some other removes.

Just run `msbuild ApiCompat.proj /p:BaselineApiCompat=true` in `src\shims`

**Comment by krwq** [comment]
@bartonjs I'm seeing couple more (except for this issue and AlgorithmIdentifier.Parameters):

```
TypesMustExist : Type 'System.Security.Cryptography.DataProtector' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.DpapiDataProtector' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.MemoryProtectionScope' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.ProtectedMemory' does not exist in the implementation but it does exist in the contract.
MembersMustExist : Member 'System.Security.Cryptography.Pkcs.EnvelopedCms..ctor(System.Security.Cryptography.Pkcs.SubjectIdentifierType, System.Security.Cryptography.Pkcs.ContentInfo)' does not exist in the implementation but it does exist in the contract.
MembersMustExist : Member 'System.Security.Cryptography.Pkcs.EnvelopedCms..ctor(System.Security.Cryptography.Pkcs.SubjectIdentifierType, System.Security.Cryptography.Pkcs.ContentInfo, System.Security.Cryptography.Pkcs.AlgorithmIdentifier)' does not exist in the implementation but it does exist in the contract.
MembersMustExist : Member 'System.Security.Cryptography.Pkcs.EnvelopedCms.Encrypt()' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.Pkcs.KeyAgreeKeyChoice' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.X509Certificates.X509Certificate2UI' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.X509Certificates.X509SelectionFlag' does not exist in the implementation but it does exist in the contract.
```

**Comment by bartonjs** [comment]
> TypesMustExist : Type 'System.Security.Cryptography.DataProtector' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.DpapiDataProtector' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.MemoryProtectionScope' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.ProtectedMemory' does not exist in the implementation but it does exist in the contract.

Not related to CMS.  When those weren't declared part of .NET Standard 2.0 we stuck with "no adding them to Core because they're Windows-only".

> MembersMustExist : Member 'System.Security.Cryptography.Pkcs.EnvelopedCms..ctor(System.Security.Cryptography.Pkcs.SubjectIdentifierType, System.Security.Cryptography.Pkcs.ContentInfo)' does not exist in the implementation but it does exist in the contract.
MembersMustExist : Member 'System.Security.Cryptography.Pkcs.EnvelopedCms..ctor(System.Security.Cryptography.Pkcs.SubjectIdentifierType, System.Security.Cryptography.Pkcs.ContentInfo, System.Security.Cryptography.Pkcs.AlgorithmIdentifier)' does not exist in the implementation but it does exist in the contract.

Seem like reasonable things to add.  They probably don't do much, but no reason to break API compat.

> MembersMustExist : Member 'System.Security.Cryptography.Pkcs.EnvelopedCms.Encrypt()' does not exist in the implementation but it does exist in the contract.

super easy.

```C#
public void Encrypt()
{
    throw new PlatformNotSupportedException(SR.Cryptography_Cms_NoRecipients);
}
```

> TypesMustExist : Type 'System.Security.Cryptography.Pkcs.KeyAgreeKeyChoice' does not exist in the implementation but it does exist in the contract.

I don't see any usage of that type in NetFX.  I wonder if it existing was an accident.  No reason to bring it to Core right now.

> TypesMustExist : Type 'System.Security.Cryptography.X509Certificates.X509Certificate2UI' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Security.Cryptography.X509Certificates.X509SelectionFlag' does not exist in the implementation but it does exist in the contract.

Not related to CMS.  And they're UI types, we don't have UI.

So a reasonable fix for all the TypesMustExist that I just said no to would be to put them in the other suppressions file
